File: solutions/python/swift-scheduling/1/swift_scheduling.py
import re
from datetime import datetime, timedelta
import calendar


def _get_eow_data(start):
    day_of_week = start.weekday()

    if 0 <= day_of_week <= 2:
        days_to_add = 4 - day_of_week
        hour = 17
    else:
        days_to_add = 6 - day_of_week
        hour = 20

    result = start + timedelta(days=days_to_add)
    return result.replace(hour=hour, minute=0, second=0, microsecond=0)


def _first_work_day_of_month(first_day):
    if first_day.weekday() == 5:
        return first_day + timedelta(days=2)
    elif first_day.weekday() == 6:
        return first_day + timedelta(days=1)
    return first_day


def _last_work_day_of_month(last_day):
    if last_day.weekday() == 5:
        return last_day - timedelta(days=1)
    elif last_day.weekday() == 6:
        return last_day - timedelta(days=2)
    return last_day


def _resolved_monthly_delivery(start, target_month):
    delivery_year = start.year if start.month < target_month else start.year + 1
    first_of_month = datetime(delivery_year, target_month, 1)
    first_work_day = _first_work_day_of_month(first_of_month)
    return first_work_day.replace(hour=8, minute=0, second=0, microsecond=0)


def _resolved_quarterly_delivery(start, target_quarter):
    current_quarter = (start.month - 1) // 3 + 1
    delivery_year = start.year + 1 if current_quarter > target_quarter else start.year

    end_month = target_quarter * 3

    last_day = calendar.monthrange(delivery_year, end_month)[1]
    last_of_quarter = datetime(delivery_year, end_month, last_day)

    last_work_day = _last_work_day_of_month(last_of_quarter)
    return last_work_day.replace(hour=8, minute=0, second=0, microsecond=0)


def delivery_date(start, description):
    normalize_date = datetime.fromisoformat(start)

    if description == "NOW":
        return (normalize_date + timedelta(hours=2)).isoformat()

    elif description == "ASAP":
        if normalize_date.hour < 13:
            scheduling_meeting = normalize_date.replace(
                hour=17, minute=0, second=0, microsecond=0
            )
        else:
            scheduling_meeting = (normalize_date + timedelta(days=1)).replace(
                hour=13, minute=0, second=0, microsecond=0
            )

    elif description == "EOW":
        scheduling_meeting = _get_eow_data(normalize_date)

    elif re.match(r"^\d{1,2}M$", description):
        target_month = int(description.replace("M", ""))
        scheduling_meeting = _resolved_monthly_delivery(normalize_date, target_month)

    elif re.match(r"^Q[1-4]$", description):
        target_quarter = int(description[1])
        scheduling_meeting = _resolved_quarterly_delivery(normalize_date, target_quarter)

    else:
        scheduling_meeting = datetime.now()

    return scheduling_meeting.isoformat()

File: solutions/java/swift-scheduling/1/src/main/java/SwiftScheduling.java
import java.time.*;

public class SwiftScheduling {
    public static LocalDateTime convertToDeliveryDate(LocalDateTime meetingStart, String description) {
        LocalDateTime schedulingMeeting;

        if (description.equals("NOW")) {
            return meetingStart.plusHours(2L);
        } else if (description.equals("ASAP")) {
            if (isASAPBefore13(meetingStart)) {
                schedulingMeeting = meetingStart.withHour(17).withMinute(0).withSecond(0).withNano(0);
            } else {
                schedulingMeeting = meetingStart.plusDays(1).withHour(13).withMinute(0).withSecond(0).withNano(0);
            }
        } else if (description.equals("EOW")) {
            int[] eowData = getEOWData(meetingStart);
            schedulingMeeting = LocalDateTime.of(eowData[0], eowData[1], eowData[2], eowData[3], 0);
        } else if (description.matches("^\\d{1,2}M$")) {
            int targetMonth = Integer.parseInt(description.replace("M", ""));
            schedulingMeeting = resolveMonthlyDelivery(meetingStart, targetMonth);
        } else if (description.matches("^Q[1-4]$")) {
            int targetQuarter = Integer.parseInt(description.substring(1));
            schedulingMeeting = resolveQuarterlyDelivery(meetingStart, targetQuarter);
        } else {
            schedulingMeeting = LocalDateTime.now();
        }

        return schedulingMeeting;
    }

    private static LocalDateTime resolveMonthlyDelivery(LocalDateTime meetingStart, int targetMonth) {
        int deliveryYear = (meetingStart.getMonthValue() < targetMonth) ? meetingStart.getYear() : meetingStart.getYear() + 1;
        LocalDate firstOfMonth = LocalDate.of(deliveryYear, targetMonth, 1);
        LocalDate firstWorkday = getFirstWorkdayOfMonth(firstOfMonth);
        return firstWorkday.atTime(8, 0);
    }

    private static LocalDateTime resolveQuarterlyDelivery(LocalDateTime meetingStart, int targetQuarter) {
        int currentQuarter = (meetingStart.getMonthValue() - 1) / 3 + 1;
        int deliveryYear = (currentQuarter > targetQuarter) ? meetingStart.getYear() + 1 : meetingStart.getYear();

        Month endMonth = Month.of(targetQuarter * 3);
        YearMonth ym = YearMonth.of(deliveryYear, endMonth);
        LocalDate lastOfQuarter = ym.atEndOfMonth();
        LocalDate lastWorkday = getLastWorkdayOfMonth(lastOfQuarter);
        return lastWorkday.atTime(8, 0);
    }

    private static LocalDate getFirstWorkdayOfMonth(LocalDate firstDay) {
        DayOfWeek dow = firstDay.getDayOfWeek();
        if (dow == DayOfWeek.SATURDAY) {
            return firstDay.plusDays(2);
        } else if (dow == DayOfWeek.SUNDAY) {
            return firstDay.plusDays(1);
        }
        return firstDay;
    }

    private static LocalDate getLastWorkdayOfMonth(LocalDate lastDay) {
        DayOfWeek dow = lastDay.getDayOfWeek();
        if (dow == DayOfWeek.SATURDAY) {
            return lastDay.minusDays(1);
        } else if (dow == DayOfWeek.SUNDAY) {
            return lastDay.minusDays(2);
        }
        return lastDay;
    }

    private static int[] getEOWData(LocalDateTime meetingStart) {
        int dayOfWeek = meetingStart.getDayOfWeek().getValue();
        LocalDateTime adjustedDate;

        if (dayOfWeek >= 1 && dayOfWeek <= 3) {
            adjustedDate = meetingStart.plusDays(5L - dayOfWeek).withHour(17).withMinute(0);
        } else {
            adjustedDate = meetingStart.plusDays(7L - dayOfWeek).withHour(20).withMinute(0);
        }

        return new int[]{
                adjustedDate.getYear(),
                adjustedDate.getMonthValue(),
                adjustedDate.getDayOfMonth(),
                adjustedDate.getHour()
        };
    }

    public static boolean isASAPBefore13(LocalDateTime meeting) {
        return meeting.getHour() < 13;
    }
}